### PR TITLE
Bug fix in EMTF GEM and RPC unpacker blocks

### PR DESCRIPTION
**Pull request by eyigitba**
#### PR description:

This PR fixes a bug that was mentioned in https://github.com/cms-sw/cmssw/issues/39456

The bug was introduced in https://github.com/cms-sw/cmssw/pull/39388

It was a stupid mistake on my part that somehow went in. The index for the for loop does not match the if/else statements, hence causing an empty GEM/RPC TP to be created which then causes the crash when we try to create detID.

This should be backported to 12_4_X and 12_5_X to fix the same ug in those releases. 

#### PR validation:

Tested with running the unpacker on the data files mentioned in https://github.com/cms-sw/cmssw/issues/39456. Unpacker runs and assigns the correct values to GEM/RPC TPs.

File used: `/eos/cms/store/t0streamer/Data/Express/000/359/045/run359045_ls0144_streamExpress_StorageManager.dat`

### CODE REVIEW

**Comment by eyigitba** [comment]
type bugfix

**Comment by eyigitba** [comment]
urgent

**Comment by tvami** [comment]
@cmsbuild , please test
 * not like anything would test this (otherwise we would have noticed it in IB tests)

**Comment by cecilecaillol** [comment]
+l1

**Comment by perrotta** [comment]
+1
- The fix is quite logical, and as we realized there is no way to actually test it with the IB tests.
- @francescobrivio is currently testing it on a previously failing offline job, see https://github.com/cms-sw/cmssw/issues/39456#issuecomment-1252884094
- Let therefore merge this in the master in order to have it included in the 2300 IB: if there are no issues, tomorrow we can also merge the backports

**Comment by perrotta** [comment]
merge

**Comment by davidlange6** [comment]
> On Sep 20, 2022, at 10:43 PM, Andrea Perrotta ***@***.***> wrote:
> 
> 
> +1
> 
> 	• The fix is quite logical, and as we realized there is no way to actually test it with the IB tests.

Why is there "no way"? (There is now clearly as express (at least) crashes) - but even in that case one can ask the requestor how was the code checked/validated.. I guess it wasn't in this case.

**Comment by eyigitba** [comment]
@davidlange6 , I verified the fix by testing it using the offline recipe as described in  https://github.com/cms-sw/cmssw/issues/39456.

Previously this error was missed in tests because I tested it using a run where GEM was offline (I wasn't aware). Since the firmware version change is quite recent, probably there were no IB tests using data collected with this firmware version. So the error did not appear in those tests.

**Comment by davidlange6** [comment]
> On Sep 21, 2022, at 9:21 AM, eyigitba ***@***.***> wrote:
> 
> 
> @davidlange6 , I verified the fix by testing it using the offline recipe as described in #39456.
> 
> Previously this error was missed in tests because I tested it using a run where GEM was offline (I wasn't aware). Since the firmware version change is quite recent, probably there were no IB tests using data collected with this firmware version. So the error did not appear in those tests.
> 

But nor did that test show that your code actually worked - thats part of my point:) [the integration procedure needs to include positive checks not just lack of negative ones..imho]

**Comment by perrotta** [comment]
> Why is there "no way"? (There is now clearly as express (at least) crashes) - but even in that case one can ask the requestor how was the code checked/validated.. I guess it wasn't in this case.

Hi David. Let me rephrase in hopefully some better English.
"The PR tests were not able to spot this problem in origin, and therefore I did not wait for them being concluded for mergin this PR, but I rather relied on the results of the dedicated tests able to reproduce the ones that were failing without this PR, see https://github.com/cms-sw/cmssw/issues/39456#issuecomment-1252902748 and https://github.com/cms-sw/cmssw/issues/39456#issuecomment-1252920127"

That said, yes I fully agree that the PR and IB tests should include Run3 data samples from runs that also include GEM, as it is not the case for the current ones. I'd let @cms-sw/pdmv-l2 deal with it and find the appropriate input files for them.

**Comment by eyigitba** [comment]
@davidlange6 , right that's a good point. I only verified the RPC part in my test and I didn't check the GEM part. On top of that, apparently I made a mistake when cleaning the RPC part which ended up being broken as well. I agree that checking for positive outcome is required (which I, apparently, failed to do properly).
